### ISSUE TITLE
Implementing CSLA - A few newbie questions

### ISSUE THREAD

**Issue by ehcanadianCA** [MarimerLLC/cslaforum#175]
I have been looking at implementing CSLA for a client app I have written and I have a few questions I was hoping I could have answered.

My business objects are currently serialized using the XMLSerializer and stored to a column in a database. That is the only storage mechanism I have available. I am open to suggestions as far as storing them in a different format, however.

Is there an issue with pulling my BO's from the database in this manner and mapping them somehow to CSLA objects? Can this be my datastore? Is there a preferred method of doing this?

The objects in the ActionExtenderSample project look like they were generated by a utility. What was used?

Thanks!

**Comment by ajj3085** [comment]
Storing your business objects like this won't end up well.  The serialization will break / become more complicated as your business objects evolve.  If you have a database, I don't see why you couldn't create proper tables to store the business objects.  Remember that BOs may share the same data, so design your DB as you normally would.  I'm not familiar with the ActionExtenderSample, but sometimes CodeSmith templates are used to generate your BOs from database tables, but if you do that, it should only be as a starting point.  You should make sure that each property really needs to be on the BO for your use case, and you should plan to reuse BOs on the behavior they provide, not the data they contain.

**Comment by ehcanadianCA** [comment]
I normally would create/use a database, but I don't have that option. I only have a text column available for storing my objects.

Does CSLA and protobuf.net play well together?

**Comment by rockfordlhotka** [comment]
Just because you aren't using a RDBMS doesn't mean you need to alter your overall architecture - it just means you need a different DAL implementation.

You can use any storage mechanism you'd like - you just can't directly serialize a CSLA object graph unless you use a full-fidelity serializer (one of: BinaryFormatter, NetDataContractSerializer, MobileFormatter).

_Really_ what you should do is run the data portal in local mode (which it does by default), and in your DataPortal_Fetch/Update/Create/Delete methods you should put the business object graph's data into a DTO object graph. Then you can serialize/deserialize that DTO graph into a file using json/xml/etc.

Basically implement your data access layer (DAL) code as normal, but using simple serialization and files instead of a real database.

**Comment by rockfordlhotka** [comment]
I don't know if it will work with protobuf.net. It depends on whether that serializer is full-fidelity like the three I listed in my previous post. If it is full-fidelity (basically has the same functionality as BinaryFormatter), then it should work fine.

But if you want to go down _that_ road, you should just use MobileFormatter, as it is part of CSLA (in the `Csla.Serialization` namespace).

**Comment by ehcanadianCA** [comment]
Thanks for the info. Is doing what you describe detailed in your book?

**Comment by rockfordlhotka** [comment]
No, the book talks about how CSLA is persistence neutral, but only includes code demonstrating how to interact with a database via Entity Framework and a mock in-memory database using LINQ.

However, it does show how to use data transfer objects (DTOs) to abstract communication between the business layer and DAL. Really all we're talking about here is changing the DAL shown in the book to directly serialize/store and then retrieve/deserialize the DTO - storing it as a string (or whatever) blob in your database column.

So in that regard yes, I suppose the book shows a lot of the required code - just replace the use of EF with serialization/deserialization of the DTO and storing it in your column instead of a relational model.

**Comment by ehcanadianCA** [comment]
An example is good enough, thanks! I appreciate you taking the time to reply.

CSLA looks great. A lot more than I need but better to be looking at it than looking for it is what I always say.

Thanks again!
